biogeophys/FatesPlantRespPhotosynthMod.F90

```diff
@@ -1214,17 +1268,75 @@ subroutine LeafLayerPhotosynthesis(f_sun_lsl,         &  ! in
               end if
-
               if (abs(gs_mol-gs_mol_err) > 1.e-01_r8) then
                  write (fates_log(),*) 'Stomatal model error check - stomatal conductance error:'
                  write (fates_log(),*) gs_mol, gs_mol_err
               end if
               
            enddo !sunsha loop
 
-           ! This is the stomatal resistance of the leaf layer
-           rstoma_out = 1._r8/gstoma
-	   
+            ! This is the stomatal resistance of the leaf layer of a given cohort
+
+            ! Adjusting gs (compute a virtual gs) that will be passed to host model
+            ! compute specific humidity from vapor pressure
+            ! q = 0.622*e/(can_press - 0.378*e) 
+            ! source https://cran.r-project.org/web/packages/humidity/vignettes/humidity-measures.html
+
+           if (hlm_use_planthydro.eq.itrue) then
+
+             !Calculates inner leaf humidity as a function of mesophyll water potential 
+             ! Adopted from  Vesala et la., 2017 https://www.frontiersin.org/articles/10.3389/fpls.2017.00054/full
+             ! LWP_star = wi/w0 = exp(psi_meso*Vh20/(R*T)) , w is the water vapor concentration 
+             ! Vh20:  molar volume of water (18 × 10−6 m3 mol−1) 
+             ! R is the universal gas constant and T the interfacial temperature 
+             ! 
+             ! th_sat = EDPftvarcon_inst%hydr_thetas_node(ft,1)
+             ! th_rs = EDPftvarcon_inst%hydr_resid_node(ft,1)
+             
+             k_lwp = EDPftvarcon_inst%hydr_k_lwp(ft)
+             if (psi<0) then
+               LWP_star = exp(((k_lwp) * psi)*18.0_r8/(8.314_r8*veg_tempk))
+               ! k_lwp**btran
+             else 
+               LWP_star = 1
+             end if
+             ! now adjust veg_esat by LWP_star
+             veg_esat_adj = veg_esat*LWP_star
+
+             ! note: q is the specific humidity
+             qs = 0.622 * ceair / (can_press - 0.378 * ceair)
+             qsat = 0.622 * veg_esat / (can_press - 0.378 * veg_esat)
+             ! qsat_adj = 0.622 * veg_esat_adj / (can_press - 0.378 * veg_esat_adj)
+             qsat_adj = qsat*LWP_star
+             
+             !  write (fates_log(),*) 'qsat:', qsat, 'RHleaf:', qsat_adj/qsat, 'RHs:', qs/qsat
+             if (k_lwp == 0 ) then 
+               rstoma_out = 1._r8/gstoma
+             else 
+               if ((qsat_adj - qs) <= 0) then
+               !  write (fates_log(),*) 'LWP_star:', LWP_star
+               !  write (fates_log(),*) 'qleaf:', qsat * LWP_star, 'qs',qs               
+                 ! if inner leaf vapor pressure is less then or equal to that at leaf surface
+                 ! then set stomata resistance to be very large to stop the transpiration or back flow of vapor
+                  rstoma_out = rsmax0*100
+               else
+                  rstoma_out = (qsat-qs)*( 1/gstoma + rb)/(qsat_adj - qs)-rb
+                
+               end if
+               if (rstoma_out <=0 ) then
```

Comment:
Do you want to keep this check?

Reply:
logically, this rstoma_out would never be negative given the formula above, but I think it is be good to keep this check just in case.